libevmasm/Instruction.h

```diff
@@ -281,19 +281,22 @@ inline Instruction logInstruction(unsigned _number)
 	return Instruction(unsigned(Instruction::LOG0) + _number);
 }
 
+/// Gas price tiers representing static cost of an instruction.
+/// Opcodes whose cost is dynamic or depends on EVM version should use the `Special` tier and need
+/// dedicated logic in GasMeter (especially in estimateMax()).
+/// The tiers loosely follow opcode groups originally defined in the Yellow Paper.
 enum class Tier
 {
-	Zero = 0,	// 0, Zero
-	Base,		// 2, Quick
-	VeryLow,	// 3, Fastest
-	Low,		// 5, Fast
-	Mid,		// 8, Mid
-	High,		// 10, Slow
-	Ext,		// 20, Ext
-	ExtCode,	// 700, Extcode
-	Balance,	// 400, Balance
-	Special,	// multiparam or otherwise special
-	Invalid		// Invalid.
+	// NOTE: Tiers should be ordered by cost, since we sometimes perform comparisons between them.
```

Comment:
We do that in https://github.com/ethereum/solidity/blob/c78f9652f5fa7c342166db527e1a8d1b0f041580/libyul/optimiser/Metrics.cpp#L185-L194

It's also the only place other than `GasMeter::runGas()` where we refer to these tiers directly. All other uses are through `GasMeter` so the removal of the obsolete tiers will have no real effect on compiler behavior.

Reply:
BTW, I didn't have time to dig into this, but I wonder if this logic is even correct. Is `Tier::BlockHash` (formerly called `Tier::Ext`) really supposed add 49 to to cost even though `BLOCKHASH` costs 20 gas? Should `TLOAD`/`TSTORE` also get 49? And isn't this also overpricing some instructions that are in `Special` not due to dynamic cost but rather becuse their cost is EVM-version dependent?

Reply:
Maybe when `Tier::Ext` was introduced the logic was not updated accordingly and it went unnoticed...
I have no idea where the value `49` comes from, but it would make sense that `Tier::BlockHash` should get a value of `20` and then the upcoming `Tier::WarmAccess` for transient storage opcodes should get `100`.
However, I also noticed that `SSTORE`, for example is in `Tier::Special`, which would make it underpriced ?